File: core/internal/sync/jump_uniqueness_test.go
package sync

import (
	"testing"

	"troubastack/core/internal/domain"
)

// P206 uniqueness, server side. Fable's ⟨D⟩ ruling makes this a LEGIBILITY rule: two jumps wearing the
// same glyph and colour corrupt nothing (pairing is by uuid, the bake resolves both), so nothing
// downstream can ever notice. The harm is to a musician meeting two identical segnos on one page mid-song.
// Studio excludes the combination in its palette; this is the refusal for what gets past it — an older
// client, a script, or two authors racing — because a rule only the UI enforces is not enforced.

// jumpEngine serves a HEAD of jump sources plus a layer→file map, which is all jumpDuplicate reads.
type jumpEngine struct {
	objects []domain.Object
	files   map[string]string // layerID → fileID
}

func (e jumpEngine) Apply(string, domain.Mutation) (domain.Mutation, error) {
	return domain.Mutation{}, nil
}
func (e jumpEngine) Head(string) (domain.Snapshot, error) {
	return domain.Snapshot{Objects: e.objects}, nil
}
func (e jumpEngine) Layer(_ string, layerID string) (domain.Layer, bool) {
	f, ok := e.files[layerID]
	return domain.Layer{ID: layerID, FileID: f, OwnerID: "me", Access: domain.AccessRW}, ok
}
func (e jumpEngine) ObjectLayer(songID, uuid string) (domain.Layer, bool, bool) {
	for _, o := range e.objects {
		if o.UUID == uuid {
			l, ok := e.Layer(songID, o.LayerID)
			return l, ok, true
		}
	}
	return domain.Layer{}, false, false
}

func jump(uuid, layerID, glyph, color, target string) domain.Object {
	return domain.Object{
		UUID: uuid, LayerID: layerID, Type: domain.TypeIcon, Text: glyph, JumpTo: target,
		Style: domain.Style{Color: color, Opacity: 1},
	}
}

func TestJumpDuplicate_PerFileGlyphAndColour(t *testing.T) {
	// One existing pair on file A (layer LA) wearing a red segno.
	existing := jump("src-1", "LA", "segno", "#e11d48", "dest-1")
	eng := jumpEngine{
		objects: []domain.Object{existing, jump("dest-1", "LA", "segno", "#e11d48", "")},
		files:   map[string]string{"LA": "fileA", "LB": "fileB"},
	}

	cases := []struct {
		name string
		obj  domain.Object
		want bool
	}{
		{"a second red segno jump on the SAME file → refused", jump("src-2", "LA", "segno", "#e11d48", "dest-2"), true},
		{"the same combination on ANOTHER file is fine — they are never read side by side (⟨D2⟩)",
			jump("src-2", "LB", "segno", "#e11d48", "dest-2"), false},
		{"a different glyph, same colour", jump("src-2", "LA", "coda", "#e11d48", "dest-2"), false},
		{"the same glyph in a different colour — the rule is the COMBINATION", jump("src-2", "LA", "segno", "#2563eb", "dest-2"), false},
		{"colour case does not make it a different look", jump("src-2", "LA", "segno", "#E11D48", "dest-2"), true},
		{"a plain landmark that only LOOKS the same is not a jump", jump("plain", "LA", "segno", "#e11d48", ""), false},
		{"re-saving the SAME jump (a move) is not a duplicate of itself", existing, false},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			c := &conn{authorID: "me", role: "member", songID: "song", hub: &Hub{eng: eng}}
			if got := c.jumpDuplicate(tc.obj); got != tc.want {
				t.Fatalf("jumpDuplicate = %v, want %v", got, tc.want)
			}
		})
	}
}

// The refusal has to reach the wire as its own reason, or the studio shows "out of date" for something a
// retry can never fix — the author needs to know to pick another glyph or colour.
func TestAuthorizeWrite_RefusesADuplicateJumpOnCreateAndOnEdit(t *testing.T) {
	eng := jumpEngine{
		objects: []domain.Object{jump("src-1", "LA", "segno", "#e11d48", "dest-1")},
		files:   map[string]string{"LA": "fileA"},
	}
	dup := objectJSON{
		UUID: "src-2", LayerID: "LA", Type: "icon", Text: "segno", JumpTo: "dest-2",
		Style: styleJSON{Color: "#e11d48", Opacity: 1},
	}
	c := &conn{authorID: "me", role: "member", songID: "song", hub: &Hub{eng: eng}}

	if reason, ok := c.authorizeWrite(domain.KindCreate, mutationJSON{UUID: "src-2", Object: &dup}); ok || reason != "jump-duplicate" {
		t.Fatalf("create of a duplicate jump = (%q, %v), want (\"jump-duplicate\", false)", reason, ok)
	}
	// An EDIT can create the collision too — recolour an existing jump onto one that already exists.
	if reason, ok := c.authorizeWrite(domain.KindSetStyle, mutationJSON{UUID: "src-1", Object: &dup}); ok || reason != "jump-duplicate" {
		t.Fatalf("edit into a duplicate = (%q, %v), want (\"jump-duplicate\", false)", reason, ok)
	}
	// The ordinary case still passes the gate.
	free := dup
	free.UUID, free.Text = "src-3", "coda"
	if reason, ok := c.authorizeWrite(domain.KindCreate, mutationJSON{UUID: "src-3", Object: &free}); !ok || reason != "" {
		t.Fatalf("a free combination = (%q, %v), want (\"\", true)", reason, ok)
	}
}

File: web/studio/test/jump-uniqueness.test.ts
import { describe, it, expect } from "vitest";
import { jumpKey, takenJumpKeys, withJumpPartners } from "../src/editor";
import type { AnnotationObject } from "../src/api";

// P206 uniqueness — Fable's ⟨D⟩ ruling: "it is a LEGIBILITY rule, not an integrity one", enforced PER FILE.
// Pairing is by uuid, so a collision corrupts nothing and no bake test can ever see it; the damage is to a
// musician reading two identical segnos on one page at a page turn. That is why it is excluded at
// authoring, and why these vectors are about what the PALETTE may offer.

function mark(uuid: string, over: Partial<AnnotationObject> = {}): AnnotationObject {
  return {
    uuid,
    layerId: "L1",
    type: "icon",
    points: [
      { x: 0.1, y: 0.1 },
      { x: 0.18, y: 0.16 },
    ],
    page: 0,
    text: "segno",
    order: 0,
    createdAt: 0,
    style: { color: "#e11d48", opacity: 1, width: 0.004, fontSize: 16 },
    ...over,
  };
}

describe("takenJumpKeys (P206 uniqueness)", () => {
  it("claims the combination for BOTH ends of a pair", () => {
    const objs = [mark("dest"), mark("src", { jumpTo: "dest" })];
    expect([...takenJumpKeys(objs)]).toEqual([jumpKey("segno", "#e11d48")]);
  });

  it("does not claim a plain cue stamp that happens to look the same", () => {
    // An ordinary icon is not a jump. Only a pair speaks for a combination.
    expect([...takenJumpKeys([mark("plain")])]).toEqual([]);
  });

  it("claims a destination that is placed but not yet paired", () => {
    // The half-finished chain is a claim in progress — handing the same combination out twice while the
    // author is mid-placement is exactly how two identical pairs get made.
    const objs = [mark("dest")];
    expect([...takenJumpKeys(objs, "dest")]).toEqual([jumpKey("segno", "#e11d48")]);
    expect([...takenJumpKeys(objs, null)]).toEqual([]);
  });

  it("is the COMBINATION, not the glyph: the same segno in another colour stays free", () => {
    const objs = [mark("dest"), mark("src", { jumpTo: "dest" })];
    const taken = takenJumpKeys(objs);
    expect(taken.has(jumpKey("segno", "#e11d48"))).toBe(true);
    expect(taken.has(jumpKey("segno", "#2563eb"))).toBe(false); // a different colour reads as different
    expect(taken.has(jumpKey("coda", "#e11d48"))).toBe(false); // a different glyph too
  });

  it("compares colours case-insensitively — #E11D48 and #e11d48 are one colour to a reader", () => {
    const objs = [
      mark("dest", { style: { color: "#E11D48", opacity: 1, width: 0.004, fontSize: 16 } }),
      mark("src", { jumpTo: "dest" }),
    ];
    expect(takenJumpKeys(objs).has(jumpKey("segno", "#e11d48"))).toBe(true);
  });

  it("claims the destination's own appearance even when the two ends differ", () => {
    // Nothing forces the two ends to match today. If they do differ, BOTH looks are spoken for — a third
    // mark copying either one would be the ambiguity the rule exists to prevent.
    const objs = [
      mark("dest", { text: "coda" }),
      mark("src", { jumpTo: "dest" }),
    ];
    const taken = takenJumpKeys(objs);
    expect(taken.has(jumpKey("segno", "#e11d48"))).toBe(true);
    expect(taken.has(jumpKey("coda", "#e11d48"))).toBe(true);
  });
});

// VLL, 2026-09-08: "deleting one of the jumpmark should delete both". A jump is ONE thing wearing two
// marks; grabbing either end takes the pair. This is what makes the earlier per-end delete — and the
// dangling pointer it left behind — impossible to create in Studio at all.
describe("withJumpPartners (P206 — a jump deletes as one)", () => {
  const dest = mark("dest");
  const src = mark("src", { jumpTo: "dest" });

  it("a source pulls its destination", () => {
    expect(withJumpPartners(["src"], [dest, src])).toEqual(["src", "dest"]);
  });

  it("a destination pulls the source aimed at it — either end takes the pair", () => {
    expect(withJumpPartners(["dest"], [dest, src])).toEqual(["dest", "src"]);
  });

  it("an ordinary mark brings nobody", () => {
    expect(withJumpPartners(["plain"], [mark("plain"), dest, src])).toEqual(["plain"]);
  });

  it("a partner that is not here (another part, or already gone) is simply not added", () => {
    expect(withJumpPartners(["src"], [src])).toEqual(["src"]);
  });

  it("keeps the selection's own order and never repeats an end", () => {
    // Selecting BOTH ends and deleting must not queue either one twice.
    expect(withJumpPartners(["src", "dest"], [dest, src])).toEqual(["src", "dest"]);
  });

  it("a mark pointing at itself does not pull itself in twice", () => {
    const selfish = mark("selfish", { jumpTo: "selfish" });
    expect(withJumpPartners(["selfish"], [selfish])).toEqual(["selfish"]);
  });
});
